kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -437,6 +445,34 @@ private void recordTransactionReport(
     engine.getMetricsReporters().forEach(reporter -> reporter.report(transactionReport));
   }
 
+  private Optional<CRCInfo> buildPostCommitCrcInfoIfCurrentCrcAvailable(
+      long commitAtVersion, TransactionMetricsResult metricsResult) {
+    if (isNewTable) {
+      return Optional.of(
+          new CRCInfo(
+              commitAtVersion,
+              metadata,
+              protocol,
+              metricsResult.getTotalAddFilesSizeInBytes(),
+              metricsResult.getNumAddFiles(),
+              Optional.of(txnId.toString())));
+    }
+
+    return readSnapshot
+        .getCurrentCrcInfo()
+        .filter(lastCrcInfo -> commitAtVersion == lastCrcInfo.getVersion() + 1)
```

Comment:
nit: want to add a comment like `// in the case of a conflicting txn and successful retry the readSnapshot may not be commitVersion - 1`

When I first saw this it was intuitive to me immediately why we needed to filter

Reply:
Side note, in those cases, is it possible for us to see if there was a CRC written and use that for simple crc write?

This can be a P2+ but maybe we should track it somewhere (i.e. create a github issue?)

Reply:
This could be a good optimization, filed https://github.com/delta-io/delta/issues/4177 for track

Reply:
This should all just be a part of the rebase process. When we rebase, we need to list + read all the JSONs and do various conflict checkpoing. We might as well list + read the latest CRC file, too.